// private/private.js
"use strict";

var defProp = Object.defineProperty || function(obj, name, desc) {
    // Normal property assignment is the best we can do if
    // Object.defineProperty is not available.
    obj[name] = desc.value;
};

// For functions that will be invoked using .call or .apply, we need to
// define those methods on the function objects themselves, rather than
// inheriting them from Function.prototype, so that a malicious or clumsy
// third party cannot interfere with the functionality of this module by
// redefining Function.prototype.call or .apply.
function makeSafeToCall(fun) {
    defProp(fun, "call", { value: fun.call });
    defProp(fun, "apply", { value: fun.apply });
    return fun;
}

var hasOwn = makeSafeToCall(Object.prototype.hasOwnProperty);
var numToStr = makeSafeToCall(Number.prototype.toString);
var strSlice = makeSafeToCall(String.prototype.slice);

var cloner = function(){};
var create = Object.create || function(prototype, properties) {
    cloner.prototype = prototype || null;
    var obj = new cloner;

    // The properties parameter is unused by this module, but I want this
    // shim to be as complete as possible.
    if (properties)
        for (var name in properties)
            if (hasOwn.call(properties, name))
                defProp(obj, name, properties[name]);

    return obj;
};

var rand = Math.random;
var uniqueKeys = create(null);

function makeUniqueKey() {
    // Collisions are highly unlikely, but this module is in the business
    // of making guarantees rather than safe bets.
    do var uniqueKey = strSlice.call(numToStr.call(rand(), 36), 2);
    while (hasOwn.call(uniqueKeys, uniqueKey));
    return uniqueKeys[uniqueKey] = uniqueKey;
}

// External users might find this function useful, but it is not necessary
// for the typical use of this module.
defProp(exports, "makeUniqueKey", {
    value: makeUniqueKey
});

function makeAccessor() {
    var secrets = [];
    var brand = makeUniqueKey();

    function register(object) {
        var key = secrets.length;
        defProp(object, brand, { value: key });
        secrets[key] = {
            object: object,
            value: create(null)
        };
    }

    return function(object) {
        if (!hasOwn.call(object, brand))
            register(object);

        var secret = secrets[object[brand]];
        if (secret.object === object)
            return secret.value;
    };
}

defProp(exports, "makeAccessor", {
    value: makeAccessor
});
